### PR TITLE
Install system wide certificates from pillars.

### PR DESCRIPTION
**Pull request by bergmannf**
`cert`-state will install the certificates as trust anchors.

### CODE REVIEW

**Comment by bergmannf** [comment]
[Providing self-signed CA for CaaSP](https://trello.com/c/nNv85jtT/808-providing-self-signed-ca-for-caasp)

File: salt/cert/init.sls [hunk 1]
```diff
@@ -7,3 +7,22 @@ include:
...
+#######################################
+
+# Install additional certificates that were setup in Velum by the user as
+# system-wide certificates
+
+{% set certs = salt.caasp_pillar.get('system_certificates', []) %}
+{% for cert_tuple in certs %}
+  {% set name, cert = cert_tuple %}
+
+/etc/pki/trust/anchors/{{ name }}.crt:
```
Comment (kiall) [inline salt/cert/init.sls]: Don't we need to issue an `update-ca-certificates` call whenever one of these certs changes?

File: salt/cert/init.sls [hunk 2]
```diff
@@ -7,3 +7,22 @@ include:
...
+  file.managed:
+    - makedirs: True
+    - contents: |
```
Comment (kiall) [inline salt/cert/init.sls]: We should explicitly set the user/group to root/root and mode to 644 or similar.

File: salt/cert/init.sls (hunk 1 above)
Comment (bergmannf) [reply salt/cert/init.sls]: That's true - I added an `onchanges` run, but that would run after every certificate.

If the user adds a many certificates that's really wasteful though, but I am not sure how to only run it after the loop if changes occured (like in `ansible` I would use register, but I didn't find an equivalent for salt).

Comment (inercia) [reply salt/cert/init.sls]: Does the `update-ca-certificates` really work on CaaSP? I remember we had some problems in the past with this command beacuse of the _read-only_ filesystrem...

Comment (flavio) [reply salt/cert/init.sls]: Yes it does nowadays

Comment (bergmannf) [reply salt/cert/init.sls]: @inercia I just ran it on a openstack deployment and it worked, so seems it does.

File: salt/cert/init.sls [hunk 3]
```diff
@@ -7,3 +7,29 @@ include:
...
+    - group: root
+    - mode: 644
+    - contents: |
+        {{ cert | indent(8) }}
+  cmd.run:
+    - name: update-ca-certificates
+    - onchanges:
+        - file: /etc/pki/trust/anchors/{{ name }}.crt
+
+{% endfor %}
```
Comment (inercia) [inline salt/cert/init.sls]: I think that, if you want to run the `update-ca-certificates` only once after several changes, you could achieve it with something like:

```SaltStack
{% set system_certs = salt.caasp_pillar.get('system_certificates', []) %}
{% for cert in system_certs %}
  {% set name, cert = cert['name'], cert['cert'] %}

/etc/pki/trust/anchors/{{ name }}.crt:
  file.managed:
    - makedirs: True
    - user: root
    - group: root
    - mode: 644
    - contents: |
        {{ cert | indent(8) }}
    - onchanges_in:
        - update-ca-certificates

{% endfor %}

update-ca-certificates:
  cmd.run

```

(This is not tested). This should insert a `onchanges` in the `update-ca-certificates` blocks for each certificate we write...

Comment (MaximilianMeister) [reply salt/cert/init.sls]: @ellisab has tested this scenario in https://github.com/kubic-project/salt/pull/537 and that didnt work. you probably need a new sls with what you suggested in https://github.com/kubic-project/salt/pull/537#discussion_r189196888

Comment (inercia) [reply salt/cert/init.sls]: Might be related to saltstack/salt#19592. Maybe you need to use `watch` instead of `onchanges`.

**Comment by kiall** [comment]
I'm OK with running it multiple times for now.. though I do agree we should work on improving with @inercia's suggestion if that works...

**Pull request merged**